### ISSUE TITLE
Add clouds in 3D mode

### ISSUE THREAD

**Issue by Ysurac** [Ysurac/FlightAirMap#363]
Like Flightgear, can be 3D clouds models and billboard using METAR as weather source.

**Comment by Ysurac** [comment]
Clouds are now available.
Planed:
* change clouds colors with the time of the day
* add rain (if possible)
* add snow (if possible ?)
* add lightning
* use others weather sources
* add wind support (this can move clouds)
* add hurricane support

**Comment by Ysurac** [comment]
Winds display support is now added for 2D mode. This use data from NOAA GFS model.

**Comment by tonkepen100** [comment]
Ohhhhh, awesome Ysu, it's a wonder, really.

Question: is there any way to change the color of the strokes left by a ship or aircraft ?, is currently blue, that color could change ?.

![captura2](https://user-images.githubusercontent.com/4571086/33035366-27c91d64-ce2b-11e7-9de7-067add363030.PNG)


Thank you.

the best Yau, the best

**Comment by Ysurac** [comment]
Open an other bug for that. This bug is only for weather.

**Comment by tonkepen100** [comment]
With respect to the Meteo that we observe, each quantum is updated ?.

This point is very important.

**Comment by Ysurac** [comment]
what do you mean by quantum ?
Winds are updated every 6 hours, it's using NOAA GFS model.
For clouds it's updated every hours, it's using airports METAR info for now. It's not really available for the sea, I will need to find an other source.

**Comment by tonkepen100** [comment]
Fantastic, of course, just like the meteo in the simulator, good, and we can know exactly when the change is made ?. I mean the time.

PD.- Sorry with the word quantum, I translated it badly.

**Comment by Ysurac** [comment]
At 0:15, 6:15, 12:15 and 18:15 for the wind.

**Comment by tonkepen100** [comment]
Well, then, there are two ideas that would be sensational ....

1.- Show somewhere, the time of update of the meteo ...

It can be done by saying which was the last hour that was updated, and which will be the next one. Or with a regressive clock, I do not know where it would be good to put this data.

2.- As you can see, when you put the mouse on the wind it informs you at a given point, the speed and its course. It would be impressive that in the information window when clicking on a ship, it could reflect that information.

I mean here .....

![captura](https://user-images.githubusercontent.com/4571086/33199665-7312f970-d0f1-11e7-99ae-7555342f0752.PNG)

We understand that the more information you provide, the more complete the tracker will be.

Thanks.

**Comment by tonkepen100** [comment]
In the 3D options, you do not see the weather, which you wanted, in your new SailAway project.

**Comment by Ysurac** [comment]
Winds and ocean wave are not available in 3D mode.
Clouds can't work on sea: It's using airport METAR, if airport is too far from the sea it's not available for now.

**Comment by tonkepen100** [comment]
Well remove from the options in your project said option.

Thank you.

**Comment by Ysurac** [comment]
No. This work when an airport is not too far. An other weather source will be used soon.

**Comment by tonkepen100** [comment]
Bien asi me gusta, que no se utilicen aeropuertos, para su nuevo proyecto.
de SailAway.

**Comment by tonkepen100** [comment]
Es suyo el proyecto no?.

**Comment by Ysurac** [comment]
Write in english. FlightAirMap support more source than only sailaway and some sail can be not too far from an airport.

**Comment by Geoneer** [comment]
1. how & where to activate the clouds?

**Comment by Ysurac** [comment]
You need to have METAR enabled (*Activate METAR support* in install script or *$globalMETAR* in **require/settings.php**) and METAR cycle enabled too (*Activate METAR cycle support* in install script or *$globalMETARcycle* in **require/settings.php**).

If it's enabled (and *scripts/update_db.php* was run) you should see *Display weather on 3D map* checkbox when you are in 3D mode in the sidebar. If there is an airport with METAR info not too far, you should see clouds.

I will change weather source asap to use something else than METAR.

**Comment by Ysurac** [comment]
When METAR is not available, I will try to use TCDC from NOMAD GFS gib2 files if possible...

**Comment by Geoneer** [comment]
ok. So I have to install the whole package on my own server. I was looking in the website of FlightAirMap. Is it visible there also?

**Comment by Geoneer** [comment]
mmm... can't access the realflight website anymore... :(

![afbeelding](https://user-images.githubusercontent.com/218929/33493207-95b1a818-d6bf-11e7-84a8-4cb35fdbd962.png)

**Comment by Ysurac** [comment]
Yes it's available on the demo website too. This work when you zoom enough on a selected flight and you are near an airport that give METAR info.

I increases some distance parameters...

The website should work now, I was working on it.

**Comment by Geoneer** [comment]
ok! You're working on it now :-). That's great! 

#clouds:
I was thinking about to merge Three.js & Cesium together like it's done here: 

https://cesium.com/blog/2017/10/23/integrating-cesium-with-threejs/

Then I will randomly let Three.js render some clouds above and around moutain area's like this is done in Zwift cycling app / Grand Theft Auto  / etc etc...

![afbeelding](https://user-images.githubusercontent.com/218929/33493539-abb9ddf0-d6c0-11e7-9480-7ccb773abb5a.png)

Real time clouds is very nice, but this is giving lot's of problems in a chain with lot of weak links. You allready got them I understand. So maybe also a nice advice for you to use Three with clouds?

**Comment by Ysurac** [comment]
I want "real" clouds with real weather.
But yes I can add some random clouds above moutain. I don't need threejs for that.

**Comment by Geoneer** [comment]
I can fully understand why you want 'real' clds! Also real weather conditions is very nice. I'm busy with this also, like in 3D games, but then in de CesiumJS globe with rain, snow, 3d tiles, etc etc.... #bigthings 

check my project here => https://www.facebook.com/smrtmps

(It's not live yett.)

**Comment by Ysurac** [comment]
Nice project.

My biggest problem is to get weather data using only latitude/longitude everywhere in the world (and for free). It's why I try to use GFS GRIB2 files...

**Comment by Geoneer** [comment]
yes.. I understand your situation. It's a big issue you're working on with real time visualization of real time weather situation on lat / lon locations. I do use http://api.openweathermap.org/data/2.5/ to get the current weather a lat / lon. 

But yeah.. that's not with cloud covering on the globe.

**Comment by Geoneer** [comment]
Another part I did implement is real time KML data from NASA about rain:

`try {
                var kml = 'https://trmm.gsfc.nasa.gov/trmm_rain/Events/3B42_rain_accumulation_3hr.kml';
                var options = {
                    camera: viewer.scene.camera,
                    canvas: viewer.scene.canvas,
                    proxy: new Cesium.DefaultProxy(proxyurl),
                    clampToGround: false,
                };
                var data = Cesium.KmlDataSource.load(kml, options);
                viewer.dataSources.add(data);
                //wait 2000ms to show the weather map. Hopefully the data is loaded before this timeout ends. Otherwise find another option to avoid the grey map over the globe.
            }
            catch (ex) {
                error(ex);
            }`

**Comment by Geoneer** [comment]
it's possible to show the KML data as a 'cloud' in the air above the Globe. I've used it a few weeks, but I was not satisfied with this. So maybe the KML can be 'rebuild' like a Elipsoid to create a 'raincloud' from out of this free data.... just think about it. 

I'm digging through your code now how you create the clouds. :-)

**Comment by Ysurac** [comment]
Thanks for the info, I will check if I can do something with that.
The code is in *js/map.3d.js*, It's the create_clouds function. It's not really clean but it's not finished yet.
I create some billboards and models at random points. I also change the color of clouds based on humidity. There is also a try to display rain commented, but it's really bad.

**Comment by Dusterka** [comment]
the winds do not change, every certain hour.

**Comment by Ysurac** [comment]
Winds data are updated every 6 hours

**Comment by Dusterka** [comment]
in my airflightmap it does not change me, and I have been observing for 24 hours.

**Comment by Ysurac** [comment]
True. NOAA is very very slow to generate GFS. The last run is 00:00...
I always use last available run now.

**Comment by Geoneer** [comment]
I found your code. I like the proof of concept of it very much! I will rebuild your methods in a more stable productive code. I will share it here also. 

And indeed I want to use highlight of the clouds in the way the 3D games do it. 

check : https://www.clicktorelease.com/blog/how-to-make-clouds-with-css-3d/

**Comment by Geoneer** [comment]
The above link is nice principle, but does not work inside cesium 3D scenes / canvas. The way you did it is the right way. But I just wanted to show a nice working example how to render the clouds inside a div.

**Comment by Geoneer** [comment]
check https://github.com/Geoneer/smrtmps/files/1484324/RTCloudsForGames_HarrisGDC2002.pdf

**Comment by Dusterka** [comment]
> True. NOAA is very very slow to generate GFS. The last run is 00:00...
> I always use last available run now.

Effectively, yesterday I changed when you said it, but I was attentive throughout the day, and not changed at the current time, that is, 24 hours. is rare.

**Comment by Ysurac** [comment]
wind is now updated correctly, GFS was really too slow.

**Comment by Geoneer** [comment]
did you try this for clouds:

https://github.com/AnalyticalGraphicsInc/cesium/wiki/Particle-System-Details 

I think I will use particles to create weatherstuff. Sounds ok for Cesium. So let's find out about this. I did use part of your code, but the model is not being show inside my viewer.entities. 

I used similar technologie with billboards, but this was a hell of a job. So for the weather stuff I will try the link https://github.com/AnalyticalGraphicsInc/cesium/wiki/Particle-System-Details

**Comment by Ysurac** [comment]
It's nice to make smoke, but for clouds I don't think it's usable and if it is, this will be really slow if we have many clouds.
I think billboard is the easiest for clouds, and model for some clouds. It's the way it's done in the Flightgear simulator.

**Comment by dextercai** [comment]
I am very curious about the GFS GRIB2 files... Is there any data or source code about the GRIB2 processing?

**Comment by Ysurac** [comment]
It's in ```require/class.weather.php```, I use grib2json so it's converted to a JSON file.
